appinventor/components/src/com/google/appinventor/components/runtime/util/FileCache.java

```diff
@@ -0,0 +1,105 @@
+package com.google.appinventor.components.runtime.util;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class FileCache {
+  public File cacheDir;
+  private final HashMap<String, CompletableFuture<Void>> fileMap = new HashMap<>();
+  private static final Logger LOG = Logger.getLogger(FileCache.class.getName());
+
+  /**
+   * Creates a new FileCache instance with the specified cache directory. If the directory doesn't
+   * exist, it will be created.
+   *
+   * @param cacheDir the directory to use for caching files
+   */
+  public FileCache(File cacheDir) {
+    this.cacheDir = cacheDir;
+    if (!cacheDir.exists()) {
+      cacheDir.mkdirs();
+    }
+  }
+
+  /**
+   * Registers a file for download from the specified URL to the cache. If the file doesn't exist in
+   * the cache, it will be downloaded asynchronously.
+   * 
+   * @param path the relative path within the cache directory where the file formshould be stored
+   * @param url the URL from which to download the file
+   * @return a CompletableFuture that completes when the download is finished, or immediately if the
+   *         file already exists
+   */
+  public CompletableFuture<Void> registerFile(final String path, final String url) {
+    final File file = new File(cacheDir, path);
+    if (!file.exists()) {
+      CompletableFuture<Void> future = CompletableFuture.runAsync(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            FileUtil.downloadUrlToFile(url, file.getAbsolutePath());
+            fileMap.remove(path);
```

Comment:
Note there is a potential race condition here where in theory the device could put the main thread to sleep in order to start the async call, and that finishes before the main thread resumes and adds the entry to the fileMap. You should synchronize access on fileMap to prevent this from happening.